src/transactions/TransactionMeta.h

```diff
@@ -0,0 +1,165 @@
+#pragma once
+
+// Copyright 2025 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "ledger/LedgerTxn.h"
+#include "transactions/EventManager.h"
+#include "transactions/MutableTransactionResult.h"
+#include "xdr/Stellar-ledger.h"
+#include <variant>
+
+namespace stellar
+{
+
+// The builder class for OperationMeta. This can only be instantiated by
+// TransactionMetaBuilder and accessed via `getOperationMetaBuilderAt`.
+class OperationMetaBuilder
+{
+  public:
+    // Sets all the ledger changes caused by this operation from the provided
+    // ledger transaction used for this operation.
+    void setLedgerChanges(AbstractLedgerTxn& opLtx);
+    // Sets the return value for a Soroban operation.
+    void setSorobanReturnValue(SCVal const& val);
+    // Returns the event manager for this operation.
+    OpEventManager& getEventManager();
+    // Returns the diagnostic event manager for the transaction.
+    // There is only one diagnstic event manager for the transaction, so this
+    // is just provided for convenience.
+    DiagnosticEventManager& getDiagnosticEventManager();
+
+  private:
+    friend class TransactionMetaBuilder;
+
+    OperationMetaBuilder(bool metaEnabled, OperationMeta& meta,
+                         OperationFrame const& op, uint32_t protocolVersion,
+                         Hash const& networkID, Config const& config,
+                         DiagnosticEventManager& diagnosticEventManager);
+    OperationMetaBuilder(bool metaEnabled, OperationMetaV2& meta,
+                         OperationFrame const& op, uint32_t protocolVersion,
+                         Hash const& networkID, Config const& config,
+                         DiagnosticEventManager& diagnosticEventManager);
+    bool maybeFinalizeOpEvents();
```

Comment:
A general comment regarding various builders, should we add a flag indicating things have been finalized and check them in the build path. This can prevent bugs of accidentally accessing the buffer after it's been moved out of memory.

Reply:
It's not a big deal if the buffer gets accessed (as long as you don't read from it), but that would indeed be a logical error, so I've added a check for the builders not being finalized.